File: python/findSuccessor.py
class BinaryTree:
    def __init__(self, value):
        self.value = value
        
        self.left = None
        self.right = None
        
        
    def InOrder(self, array, node, idx) -> list:
        if self.left != None:
            self.left.InOrder(array, node, idx)
        array.append(self)
        idx.idx += 1
        if self.value == node.value:
            idx.store = idx.idx
        if self.right != None:
            self.right.InOrder(array, node, idx)
        
        return array
    
    
    
class Idx:
    def __init__(self):
        self.idx = 0
        self.store = -1


# Time -> 0(N) Space -> 0(N)
def getSuccessor(tree, node):
    idx = Idx()
    array = tree.InOrder([], node, idx)
    if idx.store >= len(array) or idx.store == -1:
        return None
    return array[idx.store]


tree = BinaryTree(1)
tree.left = BinaryTree(2)
tree.left.left = BinaryTree(3)
tree.left.right = BinaryTree(4)
tree.right = BinaryTree(5)

node = BinaryTree(12)


print(getSuccessor(tree, node).value if getSuccessor(tree, node) != None else None)

File: go/FindSuccessor/findSuccessor.go
package main

import "fmt"

type BinaryTree struct {
	Value int

	Left  *BinaryTree
	Right *BinaryTree
}

type Idx struct {
	idx    int
	store int
	is_nil bool
}


// Time -> 0(N) Space -> 0(N) 
func (tree *BinaryTree) InOrder(array []*BinaryTree, idx *Idx, node *BinaryTree) []*BinaryTree {
	if tree == nil {
		return array
	}

	array = tree.Left.InOrder(array, idx, node)
	array = append(array, tree)
	idx.idx++
	if tree.Value == node.Value {
		idx.store = idx.idx
	}
	array = tree.Right.InOrder(array, idx, node)

	return array
}

func GetSucessor(tree, node *BinaryTree) *BinaryTree {
	idx := &Idx{store: -1}
	array := tree.InOrder([]*BinaryTree{}, idx, node)
	if idx.store >= len(array) || idx.store == -1 {return nil}
	return array[idx.store]
}


func main() {
	tree := &BinaryTree{Value: 1}
	tree.Left = &BinaryTree{Value: 2}
	tree.Left.Right = &BinaryTree{Value: 7}
	tree.Left.Left = &BinaryTree{Value: 3}
	tree.Right = &BinaryTree{Value: 12}

	node := &BinaryTree{Value: 3}

	fmt.Println(*GetSucessor(tree, node))

}